### PR TITLE
REST test framework: add historical features from production code

### PR DESCRIPTION
**Pull request by ldematte**
Check if we can use features from prod code (legacy tests do not support them). When OK, `TestFeatureService` uses `ESRestTestCaseHistoricalFeatures` as a provider, which reads historical features extracted by modules at build time (https://github.com/elastic/elasticsearch/pull/102110).

When a historical feature has not be found, we output specific error messages for legacy/new tests

### CODE REVIEW

**Comment by ldematte** [comment]
@elasticsearchmachine rerun elasticsearch-ci/part-1

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java [hunk 1]
```diff
@@ -22,18 +23,31 @@ class TestFeatureService {
...
+    ) {
         var minNodeVersion = nodeVersions.stream().min(Version::compareTo);
         var featureData = FeatureData.createFromSpecifications(specs);
         var historicalFeatures = featureData.getHistoricalFeatures();
         var allHistoricalFeatures = historicalFeatures.lastEntry() == null ? Set.of() : historicalFeatures.lastEntry().getValue();
+
+        var errorMessage = hasHistoricalFeaturesInformation
+            ? "Check the feature has been added to the correct FeatureSpecification in the relevant module or, if this is a "
+                + "legacy feature used only in tests, to a test-only FeatureSpecification"
+            : "This test seems to run on the legacy test plugins; historical features from production code will not be available."
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:40]: I think we can remove the uncertainty here - `This test is running on the legacy test framework...`

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java [hunk 2]
```diff
@@ -22,18 +23,31 @@ class TestFeatureService {
...
+                + " You need to port the test to the new test plugins in order to use historical features from production code."
+                + " If this is a legacy feature used only in tests, you can add it to a test-only FeatureSpecification";
         this.historicalFeaturesPredicate = minNodeVersion.<Predicate<String>>map(v -> featureId -> {
-            assert allHistoricalFeatures.contains(featureId) : "Unknown historical feature " + featureId;
+            assert allHistoricalFeatures.contains(featureId)
+                : String.format(Locale.ROOT, "Unknown historical feature %s: %s", featureId, errorMessage);
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:45]: `Strings.format` does the locale for you

**Review by thecoop — APPROVED** [review]
Some comments, but looks ok

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java (hunk 2 above)
Comment (ldematte) [reply test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:45]: Thanks for letting me know, that's better!

**Pull request merged**